In [345]:
import pandas as pd
import numpy as np

In [346]:
data= pd.read_csv('/home/kavin/Kavin/sem6/ML/Lab/Ex5/Telco-Customer-Churn.csv')
data

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes           Electronic check          29.85         29.85   
1                  No               Mailed check          56.95        1889.5   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038              Yes               Mailed check          84.80        1990.5   
7039              Yes    Credit card (automatic)         103.20        7362.9   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[7043 rows x 21 columns]

In [347]:
target = data['Churn']
data = data.drop(['customerID','Churn'],axis=1)

In [348]:
data['TotalCharges'] = data['TotalCharges'].replace(' ',np.nan).astype(float)

In [349]:
obj = data.select_dtypes(include=object).columns
num = data.select_dtypes(exclude=object).columns
obj

Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod'],
      dtype='object')

In [350]:
z = list(num)[1:]
z

['tenure', 'MonthlyCharges', 'TotalCharges']

In [351]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

In [352]:
sc.fit(data[z])
data[z] = sc.transform(data[z])

In [353]:
data

gender  SeniorCitizen Partner Dependents    tenure PhoneService  \
0     Female              0     Yes         No -1.277445           No   
1       Male              0      No         No  0.066327          Yes   
2       Male              0      No         No -1.236724          Yes   
3       Male              0      No         No  0.514251           No   
4     Female              0      No         No -1.236724          Yes   
...      ...            ...     ...        ...       ...          ...   
7038    Male              0     Yes        Yes -0.340876          Yes   
7039  Female              0     Yes        Yes  1.613701          Yes   
7040  Female              0     Yes        Yes -0.870241           No   
7041    Male              1     Yes         No -1.155283          Yes   
7042    Male              0      No         No  1.369379          Yes   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL             No          Yes   
1                   No             DSL            Yes           No   
2                   No             DSL            Yes          Yes   
3     No phone service             DSL            Yes           No   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7038               Yes             DSL            Yes           No   
7039               Yes     Fiber optic             No          Yes   
7040  No phone service             DSL            Yes           No   
7041               Yes     Fiber optic             No           No   
7042                No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  
0                 Yes           Electronic check       -1.160323     -0.994194  
1                  No               Mailed check       -0.259629     -0.173740  
2                 Yes               Mailed check       -0.362660     -0.959649  
3                  No  Bank transfer (automatic)       -0.746535     -0.195248  
4                 Yes           Electronic check        0.197365     -0.940457  
...               ...                        ...             ...           ...  
7038              Yes               Mailed check        0.665992     -0.129180  
7039              Yes    Credit card (automatic)        1.277533      2.241056  
7040              Yes           Electronic check       -1.168632     -0.854514  
7041              Yes               Mailed check        0.320338     -0.872095  
7042              Yes  Bank transfer (automatic)        1.358961      2.012344  

[7043 rows x 19 columns]

In [354]:
data

gender  SeniorCitizen Partner Dependents    tenure PhoneService  \
0     Female              0     Yes         No -1.277445           No   
1       Male              0      No         No  0.066327          Yes   
2       Male              0      No         No -1.236724          Yes   
3       Male              0      No         No  0.514251           No   
4     Female              0      No         No -1.236724          Yes   
...      ...            ...     ...        ...       ...          ...   
7038    Male              0     Yes        Yes -0.340876          Yes   
7039  Female              0     Yes        Yes  1.613701          Yes   
7040  Female              0     Yes        Yes -0.870241           No   
7041    Male              1     Yes         No -1.155283          Yes   
7042    Male              0      No         No  1.369379          Yes   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL             No          Yes   
1                   No             DSL            Yes           No   
2                   No             DSL            Yes          Yes   
3     No phone service             DSL            Yes           No   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7038               Yes             DSL            Yes           No   
7039               Yes     Fiber optic             No          Yes   
7040  No phone service             DSL            Yes           No   
7041               Yes     Fiber optic             No           No   
7042                No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  
0                 Yes           Electronic check       -1.160323     -0.994194  
1                  No               Mailed check       -0.259629     -0.173740  
2                 Yes               Mailed check       -0.362660     -0.959649  
3                  No  Bank transfer (automatic)       -0.746535     -0.195248  
4                 Yes           Electronic check        0.197365     -0.940457  
...               ...                        ...             ...           ...  
7038              Yes               Mailed check        0.665992     -0.129180  
7039              Yes    Credit card (automatic)        1.277533      2.241056  
7040              Yes           Electronic check       -1.168632     -0.854514  
7041              Yes               Mailed check        0.320338     -0.872095  
7042              Yes  Bank transfer (automatic)        1.358961      2.012344  

[7043 rows x 19 columns]

In [355]:
for i in num:
    data[i].fillna(data[i].mean(), inplace=True)

In [356]:
for i in data.columns[data.isna().sum()>0]:
    data[i].fillna(None, method='ffill',inplace=True)

In [357]:
from sklearn.preprocessing import LabelEncoder
enc = LabelEncoder()

for i in data.select_dtypes(include=object):
        data[i]= enc.fit_transform(data[i])

target = enc.fit_transform(target)

In [358]:
data.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'MonthlyCharges', 'TotalCharges'],
      dtype='object')

In [359]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

xtr, xtst, ytr, ytst = train_test_split(data, target, test_size=0.3)

lr = LogisticRegression()

In [360]:
lr.fit(xtr, ytr)
lr.score(xtr, ytr)

0.7981744421906694

In [361]:
from sklearn.neighbors import  KNeighborsClassifier
KNN = KNeighborsClassifier()

In [362]:
KNN.fit(xtr,ytr)
KNN.score(xtst,ytst)

0.7605300520586843

In [363]:
from sklearn.naive_bayes import GaussianNB
nv = GaussianNB()

In [364]:
nv.fit(xtr,ytr)
nv.score(xtst,ytst)

0.7567439659252247

In [370]:
from sklearn.tree import  DecisionTreeClassifier
d_tree = DecisionTreeClassifier()

In [366]:
d_tree.fit(xtr,ytr)
d_tree.score(xtst,ytst)


0.7397065783246569

In [368]:
from sklearn.svm import SVC

svc = SVC()

In [369]:
svc.fit(xtr,ytr)
svc.score(xtst,ytst)

0.8116422148603881

| Model         | Accuracy     |
|--------------|-----------|
| Logistic Regression | 79.8%     |
| KNN      | 76.05%  |
| Naive Bayes      | 75.6%  |
| Decision Trees      | 73.9%  |
| SVM      | 81.1%  |

#### Why is SVM better
- Non-linear decision bounda
- Maximum Margin Classifier
- Regularization

